### PR TITLE
Avoid making duplicate calls on Recommended Widget

### LINKED ISSUE #206: Widget: Duplicate API calls made when multiple instances of the widget are present
Related: #158.

As the JavaScript for the widget is loaded in the `widget()` method output, then there is an API call for each instance of the widget that is on the page i.e. two Parse.ly Recommended widgets means two API calls.

If the settings in each widget instance are different (maybe one personalized, maybe one not, or maybe sorted or boosted by different values), then it makes sense to do a request per widget.

When it's the same settings though, then the same results should be returned for all calls, which makes the duplicate requests redundant.

### PR DESCRIPTION
**Pull request by pauarge**
## Description

The Recommended Widget would make an API call for each instance of the widget in the page. That is suboptimal given the fact that the API call doesn't have parameters and the complete URL doesn't change much. For this reason, we have refactored the front-end code of the recommended widget to optimize the number of calls.

We have rewritten the code in such a way that we first generate the URL we should call for each widget and then we group the widgets with the same URL. Then, we do a call for each group and we then render every single widget, sharing the data.

## Motivation and Context

Closes https://github.com/Parsely/wp-parsely/issues/206

## How Has This Been Tested?

Place multiple widgets on the site and set them up so some use the "personalized" options and others don't. Open network tools and check that the number of requests gets reduced (especially if multiple widgets have similar configurations).

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by GaryJones** [comment]
> Place multiple widgets on the site and set them up so some use the "personalized" options and others don't. 

While we probably can't easily count network connections in the e2e tests, can we at least have an e2e test added that shows the widgets before this patch, and after this patch, are consistently rendered?

**Comment by pauarge** [comment]
@GaryJones Agreed that we should improve the E2E coverage on this front. I've added some tests that improve the current situation but don't really cover what we are changing here.

This is because we need to mock some network calls when we render the widget. I've been working on this for a bit and I will open a separate PR with the changes, given that they aren't straight forward.

File: src/class-parsely.php [hunk 1]
```diff
@@ -907,7 +907,7 @@ public static function post_has_trackable_status( $post ): bool {
 	 * @param WP_Post $post object.
 	 * @return array
 	 */
-	public function construct_parsely_metadata( array $parsely_options, WP_Post $post ): array {
+	public function construct_parsely_metadata( array $parsely_options, ?WP_Post $post ): array {
```
Comment (GaryJones) [inline src/class-parsely.php]: Why was this nullable needed?

File: src/UI/class-recommended-widget.php [hunk 2]
```diff
@@ -276,7 +276,7 @@ class="tiny-text" aria-labelledby="<?php echo esc_attr( $this->get_field_id( 'pu
 	public function update( $new_instance, $old_instance ): array {
 		$instance                        = $old_instance;
 		$instance['title']               = trim( wp_kses_post( $new_instance['title'] ) );
-		$instance['published_within']    = (int) trim( $new_instance['published_within'] );
+		$instance['published_within']    = is_int( $new_instance['published_within'] ) ? $new_instance['published_within'] : (int) trim( $new_instance['published_within'] );
```
Comment (GaryJones) [inline src/UI/class-recommended-widget.php:279]: Should there be a test that ensures we capture the logic being added here?

**Review by GaryJones — APPROVED** [review]
✅

File: src/UI/class-recommended-widget.php (hunk 2 above)
Comment (pauarge) [reply src/UI/class-recommended-widget.php:279]: Yes, we'll cover it here: https://github.com/Parsely/wp-parsely/issues/399

**Pull request merged**